src/cargo/sources/git/source.rs

```diff
@@ -67,13 +69,22 @@ fn ident(url: &Url) -> String {
         ident
     };
 
-    format!("{}-{}", ident, short_hash(&url))
+    Ok(format!("{}-{}", ident, short_hash(&url)))
 }
 
 // Some hacks and heuristics for making equivalent URLs hash the same
-pub fn canonicalize_url(url: &Url) -> Url {
+pub fn canonicalize_url(url: &Url) -> CargoResult<Url> {
     let mut url = url.clone();
 
+    if url.cannot_be_a_base() {
+        if url.scheme() != "github.com" {
+            return Err(format!("invalid url `{}`: cannot-be-a-base-URLs are not supported", url).into());
+        }
+        // it's most likely an attempt to fetch github repo
+        // https://github.com/rust-lang/cargo/issues/4394
+        url = Url::parse(&format!("https://github.com/{}", url.path())).unwrap();
```

Comment:
Oh for now I think it may be best to just return an error here, I don't think we're quite at the point where we want to enable these URLs yet (they're typically SSH urls, not git urls)